### PR TITLE
Redesigns the error text component (#335)

### PR DESCRIPTION
**Pull request by FrigoEU**
This is my initial attempt at (#335). It's really rough, especially the CSS is a bunch of magic numbers everywhere. Also the "algorithm" to decide whether to show the tooltip at the top or bottom is just a very simple heuristic, but I think it might work pretty well with very little performance impact. I haven't figured out how to do accurate text measuring so far.

I thought I'd show you and see what you think. Any comments very welcome :).

### CODE REVIEW

**Comment by keforbes** [comment]
Adding pretty pictures of this change.

Single-line error:
tooltip

Multi-line error:
large

Above line when close to bottom of window:
above

Functionally, this works beautifully, nice job!  I have some opinions about the colors though.  Is there a reason why you hardcoded that purple?  Did you take it from somewhere else in the current Oni styling?  I wonder if we could grab one of the colors from the current neovim theme to dynamically set the background color of that tooltip.  Or maybe just leave it gray as it was before.  Also, I liked how the previous implementation had a `0.8` opacity so it didn't *completely* cover the adjacent lines while still being readable.

I'm perfectly happy with the functionality of this error tooltip so we could merge it in, but I'd like to have a bit of a discussion about what color we want to give the background.  That's the only thing that stands out for me.  I'm afraid purple won't look good against all possible color schemes.

**Comment by FrigoEU** [comment]
Oh yeah I'm definitely open to changing the background. I think I got the color from somewhere in the UI, I think from the line numbers, but I suppose it's something that should be themed. I tried to look around a bit, but I couldn't immediately figure out how the css/color structure and theming support works or would work, so I just tried to make the first iteration of something and build on that.

I'm personally not a fan of opacity, since it makes the message harder to read. I like it on the small error markers on the right since they're always visible, but this one is only visible when you're on that line. It's your call though, I'll change it if you want to.

**Comment by bert88sta** [comment]
Getting this quasi-synced with the color scheme in vim is a must for the
future  I don't know if this is doable now

**Comment by keforbes** [comment]
> Getting this quasi-synced with the color scheme in vim is a must for the future.  I don't know if this is doable now.

Ok, I thought it'd be relatively easy since we're already pulling out `backgroundColor` and `foregroundColor` from Neovim.  I'm not sure which theme property we'd use for the error background but I thought we had a set of colors to pull from.


> I'm personally not a fan of opacity, since it makes the message harder to read. I like it on the small error markers on the right since they're always visible, but this one is only visible when you're on that line.

That's fair, you can just move the cursor to a different line if you care that the error popup is covering up an adjacent line.

File: browser/src/UI/components/Error.tsx [hunk 1]
```diff
@@ -78,24 +83,41 @@ export class ErrorMarker extends React.Component<IErrorMarkerProps, void> {
...
-
-        const errorDescription = this.config.getValue("editor.errors.slideOnFocus") ? (<div className="error">
-            <div className="text">
-                {this.props.text}
-            </div>
-        </div>) : null
-
-        return <div style={positionDivStyles} className={className}>
-            {errorDescription}
+        const className = [
```
Comment (extr0py) [inline browser/src/UI/components/Error.tsx]: Minor suggestion here - you can use the `classnames` (https://github.com/JedWatson/classnames) module to make this a bit easier, without needing the `join(" ")`. We already have it in the project, the 'oni\browser\src\UI\components\LiveEvalMarker.tsx` uses it

**Comment by extr0py** [comment]
I'm very sorry for the slow response on my end! Had some family obligations this weekend so wasn't on the computer much. This is an awesome PR @FrigoEU - great to see the design & usability improvements! Excellent work. 👍 👍 👍 

I really like the restyling and the thought you put in as to whether it should open above or below.

I agree with the sentiment around the background color - it would make sense for this to be consistent with the quick info, completion, and popup menu UI. We have the `@background-color` constant in .less we could leverage. Regarding the colorscheme - makes sense to sync once we are able to do that - that work is tracked by #33. But it won't be straightforward until we address that issue, so just making it match with the other elements would be a reasonable goal. 

I experimented a bit and using the `@background-color` constant along with a drop shadow (`box-shadow: 3px 3px rgba(0, 0, 0, 0.2)` doesn't look too bad and is relatively consistent:
image

The only other thing I was thinking about was - does it make sense for the errors to be centered under the item? Something like this:
image.

This is the sort of experience and change that helps move us towards the __Modern UX__. Looking at the current experience and this PR, it's night and day, and it is leaps and bounds ahead of the UI that is possible with the terminal limitations. Awesome job! 👍 👍 

The only item I'd like to see addressed is making the background color consistent with the other popup elements - otherwise I'm ready to merge it in. Can't wait to start using it! 😄

**Comment by FrigoEU** [comment]
I made the changes to use the @background-color variable and the dropshadow as mentioned above. I think it's not a bit idea to try it out for a bit and then see what everybody thinks about it. There are still issue with the errors, like when you have two on one line they go over/under eachother, so this is definitely not the end. 

Just wanted to say thanks for the enthousiasm and encouragement I've been getting on all my PR's, it makes all the difference.

**Comment by FrigoEU** [comment]
One thing that I noticed and that is really annoying is that autocompletions are positioned under the error messages. I can fix this with some z-index magic, but I think that's a pretty bad idea. As I can see, the completions get rendered inside the "overlay-ui" container which is present in the index.html so I think that's quite good, easy to find. The errors however get positioned inside the DOM by the OverlayManager, which just appends itself to the document on startup and then renders all its overlays inside, so it's always later in the DOM than the containers that are initially present in the index.html.

It hurts my head a little bit trying to figure out the proper structure to handle the correct level positioning of all these overlay elements. As I can see we have these "kinds" of overlays:

* QuickOpen / Ctrl-P
* QuickInfo
* Autocompletions
* Error markers / Errors
* Live Evaluation
* Scrollbar

Honestly, I don't think we'll have that many more overlay kinds popping up over time so maybe it's not that big a deal, but it is a bit annoying that some are managed by the OverlayManager (the bottom 3), and others are just rendered directly in the RootComponent (top 3). I'm more of a fan of keeping everything in the RootComponent hierarchy, but that's just me. I'd love some guidance on the direction we want to take here, if any.

**Comment by extr0py** [comment]
> I made the changes to use the @background-color variable and the dropshadow as mentioned above. I think it's not a bad idea to try it out for a bit and then see what everybody thinks about it. There are still issue with the errors, like when you have two on one line they go over/under eachother, so this is definitely not the end.

Sounds great! I checked out the changes and they fit in well with the UI. Thanks for making the updates. As you mentioned we can always tweak it later, but this moves much closer to the experience you get with something like Atom!

> Just wanted to say thanks for the enthousiasm and encouragement I've been getting on all my PR's, it makes all the difference.

Definitely, thank you for taking the time to learn about and contribute to the project, and for being so open to feedback! Really appreciate your contributions.

> It hurts my head a little bit trying to figure out the proper structure to handle the correct level positioning of all these overlay elements. As I can see we have these "kinds" of overlays:

Unfortunately, the overlay management system is a mess at the moment. IMO it is the most convoluted and confusing part of the codebase at the moment. It hurts my head too and I wrote it 😦 The main problem is there is a bunch of complexity in terms of managing the 'overlays' today - because the Neovim instance is one giant `<div>`, and when there is a split or changes in the split size, there are a bunch of hacky hooks to get the errors and scrollbar to render 'on top' of the appropriate area. 

I'd like to look at _handling window management explicitly_ and _multiplexing Neovim instances_ to solve this problem.  This has been alluded to in a couple PRs, but I logged this issue - #362 - to explicitly track this. 

The goal with that work would be to keep everything in our React component hierarchy, so we don't hit those annoyances you mentioned. Let me know if you want to chat more. PR #303 is the first step I had in refactoring some of the React components and creating a React component to host the Neovim process - #362 might be more doable after that PR.

**Comment by extr0py** [comment]
I'll bring this change in! Thanks for the contribution, @FrigoEU , excellent work!

**Pull request merged**